src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/ActiveDirectoryAuthenticationProvider.cs

```diff
@@ -198,6 +199,49 @@ public override async Task<SqlAuthenticationToken> AcquireTokenAsync(SqlAuthenti
                 return new SqlAuthenticationToken(accessToken.Token, accessToken.ExpiresOn);
             }
 
+            if (parameters.AuthenticationMethod == SqlAuthenticationMethod.ActiveDirectoryFederatedIdentityCredentials)
+            {
+                // We'll keep using common environment variables defined in Azure.Identity library
+                // AZURE_TENANT_ID: The tenant ID where target resource hosts
+                // AZURE_CLIENT_ID: The client id of multi-tenant entra id app. May be overridden by the User Id.
+                // AZURE_MSI_CLIENT_ID (new): The client id of managed identity
+                string tenantId = Environment.GetEnvironmentVariable("AZURE_TENANT_ID");
+                tenantId = string.IsNullOrWhiteSpace(tenantId) ? null : tenantId;
+
+                if (clientId is null)
+                {
+                    clientId = Environment.GetEnvironmentVariable("AZURE_CLIENT_ID");
+                    clientId = string.IsNullOrWhiteSpace(clientId) ? null : clientId;
+                }
+
+                string managedIdentityClientId = Environment.GetEnvironmentVariable("AZURE_MSI_CLIENT_ID");
+                managedIdentityClientId = string.IsNullOrWhiteSpace(managedIdentityClientId) ? null : managedIdentityClientId;
+
+                ClientAssertionCredential clientAssertionCredential = null;
+                // If either tenant id, client id, or the managed client id are not specified when fetching the token,
+                // a CredentialUnavailableException will be thrown instead
+                if (!string.IsNullOrEmpty(tenantId) && !string.IsNullOrEmpty(clientId) && !string.IsNullOrEmpty(managedIdentityClientId))
+                {
+                    clientAssertionCredential = new ClientAssertionCredential(
+                        tenantId,
+                        clientId,
+                        async (cancellationToken) =>
+                        {
+                            ManagedIdentityCredential miCredential = new ManagedIdentityCredential(managedIdentityClientId);
+                            var token = await miCredential.GetTokenAsync(new TokenRequestContext(new string[] { $"api://AzureADTokenExchange/.default" })).ConfigureAwait(false);
+                            return token.Token;
+                        });
+                }
+
+                if (clientAssertionCredential == null)
+                {
+                    throw new CredentialUnavailableException("FederatedIdentityCredentials authentication unavailable. The required environment variables are not fully configured.");
```

Comment:
This string will need to be moved to string resources for localization.

Reply:
Done